### ISSUE TITLE
[bug] Replacement for `scm` attribute has logical bug

### ISSUE THREAD

**Issue by gpxricky** [conan-io/conan#10971]
Hi,

I'm currently porting my Conan recipes to be prepared for Conan 2.X. I've already stumbled several times about the replacement of the `scm` feature. At the end, I think the new concept is much better.

However, I think the example for the replacement of the `scm` attribute (https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#the-scm-attribute) is logically wrong or has a major defect. The example is as follows:

```
from conan import ConanFile
from conan.tools.scm import Git
from conan.tools.files import load, update_conandata

class Pkg(Conanfile):

   def export(self):
       git = Git(self, self.recipe_folder)
       scm_url, scm_commit = git.get_url_and_commit()
       update_conandata(self, {"sources": {"commit": scm_commit, "url": scm_url}})

   def source(self):
       git = Git(self)
       sources = self.conan_data["sources"]
       git.clone(url=sources["url"], target=".")
       git.checkout(commit=sources["commit"])
```

From my experience, this works fine if the recipe folder equals the latest commit of the whole repository. However, if the recipe folder has a different commit than the entire repository, then the wrong Git information will be written to the `conandata.yml`.

I think the major problem is the implementation of the `get_commit` method of the new `Git` tool (https://github.com/conan-io/conan/blob/develop/conan/tools/scm/git.py#L18). It returns the commit of the current folder, not of the entire repository. I propose to either use a different method in `get_url_and_commit()` that fetches the overall commit or to add a parameter to the `get_commit` method that is able to control its behavior. In any case I think `get_url_and_commit()` shall return the commit of the whole repo.

### Environment Details (include every applicable attribute)
  * Conan version: 1.46.0

Best regards,
Markus

**Comment by lasote** [comment]
Hi, The second argument of the `Git` object can be set with the correct folder (the folder you want to clone). If the recipe is not the root, Would it work something like:

```python

class Pkg(ConanFile):

   def export(self):
       git = Git(self, os.path.join(self.recipe_folder, ".."))
       scm_url, scm_commit = git.get_url_and_commit()
       update_conandata(self, {"sources": {"commit": scm_commit, "url": scm_url}})

   def source(self):
       git = Git(self)
       sources = self.conan_data["sources"]
       git.clone(url=sources["url"], target=".")
       git.checkout(commit=sources["commit"])
```

Note the `git = Git(self, os.path.join(self.recipe_folder, ".."))`

**Comment by gpxricky** [comment]
@lasote 

This still doesn't work for me. Imagine the following situation which I would call a regular one:

```
git init
touch root.txt
mkdir subfolder
touch subfolder/subfolder.txt
git add .
git commit -m "Initial commit"
git checkout -b change_branch
echo "Content" > subfolder/subfolder.txt
git add .
git commit -m "Second commit"
git checkout master
git merge --no-ff change_branch -m "Merge branch 'change_branch'"
```

Now, check the latest commit of the entire repository:

```
git rev-parse HEAD
```

And then compare it to what Conan would calculate in the `get_commit` method (even with your proposal):

```
git rev-list HEAD -n 1 -- .
```

You will see the results are different.

The example has been tested with Git version 2.17.1 running on Ubuntu 18.04.

By the way: please be aware that in the example above the default branch is called "master". You need to change that to "main" or whatever your Git configuration expects.

**Comment by lasote** [comment]
I see, thanks for the example. I've assigned it to the next release to investigate it.

**Comment by DukeOfKirkcaldy** [comment]
I'm also having this issue when exporting the recipe from a merge commit (i.e. the would-be result of a merge which is stored on Github when a pull request is opened).

As you can see I am on commit 428dd13..., but Conan calculates a different :
```bash
/c/Projects/conan-demo/conan-demo-liba ((428dd13...))
$ git rev-list HEAD -n 1 -- .
884a778fa2610e0de5b8a504fd30954cf8fba197
```

It turns out the result from Conan is the previous commit.
```
/c/Projects/conan-demo/conan-demo-liba ((428dd13...))
$ git rev-list HEAD
428dd13861a57a1fc05b00b09131f033851aa33e
884a778fa2610e0de5b8a504fd30954cf8fba197
...
```

I'm not exactly sure why this is the case as the merge commit does modify files directly under the root dir, '.'...

Edit: I think I found the reason for this in the [Git documentation](https://git-scm.com/docs/git-rev-list#Documentation/git-rev-list.txt-Defaultmode):
> Default mode
> Simplifies the history to the simplest history explaining the final state of the tree. Simplest because it prunes some side branches if the end result is the same (i.e. merging branches with the same content)

**Comment by lasote** [comment]
Thanks! valuable feedback.

**Comment by lasote** [comment]
Looks like we could have a better result using `git rev-list HEAD -n 1 --full-history -- .`
I wanted to explain why insisting on this and not just use `rev-parse HEAD`, there are users having a mono-repo with N subfolders, each of them with a recipe. So when they want the revision to be the `commit` of the latest changes in the subfolder, not in all of them, because otherwise, any change in a recipe would change the revision in the others.

@GordonJess @gpxricky Could you try with the suggested command and check if that is more expected?

**Comment by DukeOfKirkcaldy** [comment]
Hi @lasote,

Thanks!  I can confirm your suggestion works in my case:

```
/c/Projects/conan-demo/conan-demo-liba ((bfd15e1...))
$ git rev-list HEAD -n 1 --full-history -- .
bfd15e19d4b42ccec803dc27f97debc620239ec7
```

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/11015 merged, will be released 1.48. Thanks!
